db/snapshot/backup_task.cc

```diff
@@ -99,52 +101,133 @@ future<> backup_task_impl::upload_component(sstring name) {
-        co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+    co_return greedy_partitioning(std::move(backup_files), smp::count);
+}
+} // namespace
 
-        std::ignore = upload_component(component_ent->name).handle_exception([&ex] (std::exception_ptr e) {
-            // keep the first exception
-            if (!ex) {
-                ex = std::move(e);
-            }
-        }).finally([gh = std::move(gh)] {});
-        co_await coroutine::maybe_yield();
-        co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
-        if (impl::_as.abort_requested()) {
-            ex = impl::_as.abort_requested_exception_ptr();
-            break;
-        }
+future<> backup_task_impl::do_backup() {
+    if (!co_await file_exists(_snapshot_dir.native())) {
+        throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
     }
 
-    co_await snapshot_dir_lister.close();
-    co_await uploads.close();
-    if (ex) {
+    auto backup_shards = co_await get_backup_files(_snapshot_dir);
+
+    if (snap_log.is_enabled(log_level::trace))
+        snap_log.trace("{}", print_results(backup_shards));
+
+    sharded<abort_source> as;
+    co_await as.start();
+    auto s = _as.subscribe([&]() noexcept {
+        try {
+            std::ignore = smp::invoke_on_all([&as, ex = _as.abort_requested_exception_ptr()] {
+                as.local().request_abort_ex(ex);
+            });
+        } catch (...) {
+        }
+    });
+
+    std::exception_ptr ex;
+    co_await smp::invoke_on_all([this, &ex, &backup_shards, &as] -> future<> {
+        try {
+            auto shard_id = this_shard_id();
+            auto cln = _snap_ctl.storage_manager().container().local().get_endpoint_client(_endpoint);
+
+            gate uploads;
+
+            for (const auto& name : backup_shards[shard_id].files) {
+                // Pre-upload break point. For testing abort in actual s3 client usage.
+                co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+                auto gh = uploads.hold();
+
+                try {
+                    co_await upload_component(cln, as.local(), _progress_per_shard[shard_id], name);
+                } catch (...) {
+                    ex = std::current_exception();
+                    break;
+                }
+                co_await coroutine::maybe_yield();
+                co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
+                if (as.local().abort_requested()) {
+                    ex = as.local().abort_requested_exception_ptr();
+                    break;
+                }
+            }
+            co_await uploads.close();
+        } catch (...) {
+            ex = std::current_exception();
```

Comment:
Not sure if it's valid to assign current_exception from one shard to a ptr variable from another

Reply:
What if I'll propagate exceptions using `_as` abort source? Looks like it is working as expected but I'm not sure it is a valid situation when I call `request_abort_ex` on `_as` from a shard that does not own the `_as`